### ISSUE TITLE
Instances are not removed from Cloud Map when deleting the virtualNode

### ISSUE THREAD

**Issue by bemer** [aws/aws-app-mesh-controller-for-k8s#107]
If you create a VirtualNode in the kubernetes with `spec.serviceDiscovery.cloudMap` the controller register the Pods IP Addresses within the Cloud Map namespace, but when you delete the virtualNode the pods are still registered in the Cloud Map namespace.

I was using this VirtualNode spec:

```
apiVersion: appmesh.k8s.aws/v1beta1
kind: VirtualNode
metadata:
  name: nodejs-app
  namespace: appmesh-workshop-ns
spec:
  meshName: appmesh-workshop
  listeners:
    - portMapping:
        port: 3000
        protocol: http
  serviceDiscovery:
    cloudMap:
      namespaceName: appmeshworkshop.pvt.local
      serviceName: nodejs
```

The way to get the instances deregistered from Cloud Map was using the Cloud Map api: 

```
SERVICE_ID=$(aws servicediscovery list-services --filters Name="NAMESPACE_ID",Values=$NAMESPACE,Condition="EQ" | jq -r ' .Services[] | [ .Id ] | @tsv ' )
aws servicediscovery list-instances --service-id $SERVICE_ID | jq -r ' .Instances[] | [ .Id ] | @tsv ' |\
  while IFS=$'\t' read -r instanceId; do 
    aws servicediscovery deregister-instance --service-id $SERVICE_ID --instance-id $instanceId
  done
```

**Comment by nckturner** [comment]
I can confirm this, it looks like we do not have logic for cleaning up the instances and the cloudmap service when the virtualnode is deleted.  Arguably not a bug, this was intended as a follow up task from the original Cloud Map support PR, but it looks like it fell through the cracks. As a side note, when pods are deleted, Individual instance deregistration does work.  

If we can know that the virtualnode is 1:1 with the Cloud Map service, then we should attempt to delete it.  Regardless, we should deregister instances from the service when the virtual node is deleted.  We do need a strategy for keeping track of these secondary resources though, that are not explicitly identified by a custom resource, but we still create.

**Comment by lavignes** [comment]
> If we can know that the virtualnode is 1:1 with the Cloud Map service

A virtual node can select a subset of instances in a CloudMap service via Attributes. So many virtual nodes can point to instances of the same service.

**Comment by nckturner** [comment]
@lavignes so in ECS, how do the instances get registered?  Service discovery isn't useful unless the endpoints are correct.

**Comment by lavignes** [comment]
@nckturner ECS registers the endpoints.

See this example: https://github.com/aws/aws-app-mesh-examples/tree/master/walkthroughs/howto-http2

Specifically we associate the registry (cloudmap) with the ECS service: https://github.com/aws/aws-app-mesh-examples/blob/master/walkthroughs/howto-http2/app.yaml#L523

ECS attaches its own metadata when this is done. For example the `AVAILABILITY_ZONE` attribute. Individual instances of the same service can/will have different AZs. By specifying a `AVAILABILITY_ZONE` attribute I can route traffic to a specific AZ in app mesh.

**Comment by kiranmeduri** [comment]
I am working on a fix to deregister instances when virtual-node is deleted (as part of finalizer logic). Regarding Cloud Map service management it is not so straight-forward. There is no way to be certain that Cloud Map service is not used within this cluster or elsewhere without proper bookkeeping. We can tag Cloud Map service with a owner token and use it to perform management. We can also delete service if there are no instances. Either ways this needs a design proposal.

**Comment by astrived** [comment]
This issue is addressed as part of #153
